### PR TITLE
Add nvCOMP & Use Hybrid Depth Compression

### PR DESCRIPTION
**Pull request by TomaszTB**
This PR adds tools for using nvCOMP. See any of the blow linked spread sheets for performance measurements. 

This PR also adds a lossless/lossy hybrid depth compression method. For 16 bit depth images, the hybrid method uses lossless compression (zstd) for the 8 most significant bits, and a lossy method (jpeg) for the 8 least significant bits. This way we can achieve high compression ratios while keeping artifacts to a low. 

| quality | compression ratio | spread sheet link |
|----|-------|-|
| 95 | 3.19  | [link](https://docs.google.com/spreadsheets/d/1JUXtFvnDTJuNvQhBU8uNALw481dmB9BwgtAUlpD_MrM/edit?usp=sharing) |
| 75 | 5.56  | [link](https://docs.google.com/spreadsheets/d/1Y2CfJaDJuEVrsOGhTYXMa9CuW4_-ouycLPfvbtn3K_E/edit?usp=sharing) |
| 40 (current default) | 7.66  | [link](https://docs.google.com/spreadsheets/d/10q_v2c3G-CJauG3dQsCymbm5EcTZpqiX_lsMZwSTJH4/edit?usp=sharing) |
|  1 | 14.89 | [link](https://docs.google.com/spreadsheets/d/1RS3gWW5TfjlxEKG0rKz_iW17HMeF_R5IRUuM80Tm5SI/edit?usp=sharing) |

CUDA is required to decompress depth images compressed using this method, meaning that this PR introduces a CUDA requirement to the `RDXROS2ImageMessageVisualizer` (only for rendering depth images) and `RDXROS2ColoredPointCloudVisualizer`.

### CODE REVIEW

File: ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/ros2/RDXROS2ImageMessageVisualizer.java [hunk 1]
```diff
@@ -128,6 +130,18 @@ private void queueRenderImage(Subscriber<ImageMessage> subscriber)
                      decompressedImage = new Mat(imageHeight, imageWidth, opencv_core.CV_16UC1);
                      normalizedScaledImage = new Mat(imageHeight, imageWidth, opencv_core.CV_32FC1);
                   }
+                  case DEPTH_HYBRID_ZSTD_JPEG_16UC1 ->
+                  {
+                     LogTools.info("Creating Image Message Visualizer for {} with type DEPTH_HYBRID_16UC1", topic.getName());
```
Comment (ds58) [inline ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/ros2/RDXROS2ImageMessageVisualizer.java]: I know you were just pattern matching here but can you do like

```
LogTools.info("Creating Image Message Visualizer for {} with type {}", topic.getName(), ImageMessageFormat.getFormat(imageMessage).name());
```

below the switch statement so you don't have to duplicate it for each case

File: ihmc-perception/src/main/java/us/ihmc/perception/demo/NVCompDemo.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+import java.util.Map.Entry;
+import java.util.TreeMap;
+
+import static org.bytedeco.cuda.global.cudart.*;
+import static org.bytedeco.cuda.global.nvcomp.*;
+import static us.ihmc.perception.cuda.CUDATools.checkCUDAError;
+
+public class NVCompDemo
+{
+   protected static final String FILE_PATH = "/home/tbialek/Documents/ZED/ZED_ColorImage.png";
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/demo/NVCompDemo.java]: Remove hard coded absolute path

File: ihmc-perception/src/main/java/us/ihmc/perception/demo/NVCompDemo.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+         System.out.println("\tCompression Time:   " + compressionTime);
+         System.out.println("\tDecompression Time: " + decompressionTime);
+
+         checkCUDAError(cudaStreamSynchronize(stream));
+         checkCUDAError(cudaFreeHost(compressedImageData));
+         decompressedImage.close();
+      }
+   }
+
+   protected long compress(Mat image, BytePointer compressedImageData, PimplManager compressionManager)
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/demo/NVCompDemo.java]: Can this be removed in favor of the CUDACompressionTools class?

File: ihmc-perception/src/main/java/us/ihmc/perception/demo/NVCompDemo.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+      checkCUDAError(cudaFreeAsync(compressedDeviceBuffer, stream));
+      checkCUDAError(cudaFreeAsync(decompressedDeviceBuffer, stream));
+      decompressedDeviceBuffer.close();
+      compressedDeviceBuffer.close();
+      compressionConfig.close();
+
+      return compressedBufferSize;
+   }
+
+   protected void decompressImage(BytePointer compressedHostBuffer, long compressedBufferSize, Mat matToPack)
```
Comment (ds58) [inline ihmc-perception/src/main/java/us/ihmc/perception/demo/NVCompDemo.java]: Can this be removed in favor of the CUDACompressionTools class?

**Review by ds58 — CHANGES_REQUESTED** [review]

**Comment by calvertdw** [comment]
Seeing this, it looks a little choppy. Would be better to see it on real robot data, however, I think the quality should be a little higher.
image

**Comment by ds58** [comment]
I think we should change the topics in this PR so it's like:
```
/ihmc/d455/image/depth/compressed
```

That way it's clear you probably shouldn't run perception algorithms on it.

**Comment by ds58** [comment]
> Seeing this, it looks a little choppy. Would be better to see it on real robot data, however, I think the quality should be a little higher. image

I think it's fairly reasonable considering that it's only for visualization. But yeah, we should test on some different environments first.

**Review by calvertdw — APPROVED** [review]

**Comment by calvertdw** [comment]
> > Seeing this, it looks a little choppy. Would be better to see it on real robot data, however, I think the quality should be a little higher. image
> 
> I think it's fairly reasonable considering that it's only for visualization. But yeah, we should test on some different environments first.

It's one thing for us to see the data, but screenshots and video recordings of our UI also represent our organization. I'd like us to look good in that respect, too. So just under the obvious level of degradation would be nice. Additionally, I think sometimes people would get confused and think the sensor sucks, and subliminally make assumptions going forward. I remember myself doing that a lot early on.

There's also the fact that through, interdependence, visualization of data is actionable in our UI, through teleoperation and behavior authoring. So in that respect, the operator is a "perception algorithm".

**Comment by TomaszTB** [comment]
> I think we should change the topics in this PR so it's like:
> 
> ```
> /ihmc/d455/image/depth/compressed
> ```
> 
> That way it's clear you probably shouldn't run perception algorithms on it.

Unfortunately the suffix is already taken by the "depth" part. Not sure how we can add more after the suffix. 

Can we make that a separate PR? I feel like it doesn't relate to nvCOMP and using a different depth compression method.

**Comment by ds58** [comment]
> > I think we should change the topics in this PR so it's like:
> > ```
> > /ihmc/d455/image/depth/compressed
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That way it's clear you probably shouldn't run perception algorithms on it.
> 
> Unfortunately the suffix is already taken by the "depth" part. Not sure how we can add more after the suffix.
> 
> Can we make that a separate PR? I feel like it doesn't relate to nvCOMP and using a different depth compression method.

What if we use `/lossy` instead of `/compressed`? We are changing that aspect where it used to be lossless and now it is lossy.

**Comment by TomaszTB** [comment]
> > > I think we should change the topics in this PR so it's like:
> > > ```
> > > /ihmc/d455/image/depth/compressed
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > That way it's clear you probably shouldn't run perception algorithms on it.
> > 
> > 
> > Unfortunately the suffix is already taken by the "depth" part. Not sure how we can add more after the suffix.
> > Can we make that a separate PR? I feel like it doesn't relate to nvCOMP and using a different depth compression method.
> 
> What if we use `/lossy` instead of `/compressed`? We are changing that aspect where it used to be lossless and now it is lossy.

Ah, that does relate more. Still will have to figure out how to add the "lossy" into there.

**Comment by TomaszTB** [comment]
> > > Seeing this, it looks a little choppy. Would be better to see it on real robot data, however, I think the quality should be a little higher. image
> > 
> > 
> > I think it's fairly reasonable considering that it's only for visualization. But yeah, we should test on some different environments first.
> 
> It's one thing for us to see the data, but screenshots and video recordings of our UI also represent our organization. I'd like us to look good in that respect, too. So just under the obvious level of degradation would be nice. Additionally, I think sometimes people would get confused and think the sensor sucks, and subliminally make assumptions going forward. I remember myself doing that a lot early on.
> 
> There's also the fact that through, interdependence, visualization of data is actionable in our UI, through teleoperation and behavior authoring. So in that respect, the operator is a "perception algorithm".

I added a link to 75 quality. I think there's minor improvement above 75. Also, the images I included are quite zoomed in. I'll try to get a video in the lab so we can see what it looks like in normal use.

**Review by ds58 — APPROVED** [review]

**Pull request merged**